### ISSUE TITLE
ResponseCaching with authorize

### ISSUE THREAD

**Issue by smartcodinghub** [aspnet/Home#2958]
Hi,

I'm trying to implement ResponseCaching but all my request are Authenticated. My exact needs are to cache all GET request varying on all the query (with *) and on the Authorization header.

I don't find a way to do this and the official page that you can't cache requests with the Authorization header. 

Since near all the applications needs some kind of authentication, I can't find a way to not support this. It can even validate the authorization but returning the cached result instead of calculating it.

**Comment by Tratcher** [comment]
Public caching if authenticated resources is highly ineffective. Such resources should only be privately cached on the client.

**Comment by smartcodinghub** [comment]
That depends on the use case. We have only a bunch of clients concurrently, but they make tons of request. So I don't find it that ineffective.

You are suggesting to only add the Cache-Control to all that requests with private?

**Comment by Tratcher** [comment]
Yes, cache user specific content privately on that user's local machine, not in a public cache on the server.

**Comment by JonasSamuelsson** [comment]
I think this can be done by registering your own implementation of `IResponseCachingPolicyProvider` in the service provider.  
Inherit from `ResponseCachingPolicyProvider` and override `AttemptResponseCaching(ResponseCachingContext)`.  

This does feel a bit dirty though as both `IResponseCachingPolicyProvider`, `ResponseCachingPolicyProvider` and `ResponseCachingContext` all live in the `Microsoft.AspNetCore.ResponseCaching.Internal` namespace.
